File: Choonghun/공통문제/12주차/프로그래머스 76503 - 모두 0으로 만들기/모두 0으로 만들기.py
import sys
sys.setrecursionlimit(10**6)
def solution(a, edges):
    answer = -1
    if sum(a):
        return answer
    answer = [0]
    stack = []
    n = len(a)
    graph = [[] for _ in range(n)]
    for i, j in edges:
        graph[i].append(j)
        graph[j].append(i)
    
    def dfs(cur, prev):
        for next in graph[cur]:
            if prev != next:
                a[cur] += dfs(next, cur)
                answer[0] += abs(a[next])
                a[next] = 0
        return a[cur]
    
    dfs(0, -1)
    return answer[0]

File: Gwanghun/공통문제/12주차/프로그래머스/3/76503. 모두 0으로 만들기/모두 0으로 만들기.cpp
#include <string>
#include <vector>

using namespace std;

vector<vector<int>> graph(300001);

bool isVisited[300001];
long long result=0;
long long total=0;

void init(vector<vector<int>> edges){
    
    for(int i=0; i<edges.size(); i++){
        int fir=edges[i][0];
        int sec=edges[i][1];
                
        graph[fir].push_back(sec);
        graph[sec].push_back(fir);
    }    
}

long long dfs(vector<int>& a, int curVal) {
    
    isVisited[curVal]=true;
    
    long long sum=a[curVal];
    
    for(int i=0; i<graph[curVal].size(); i++){
        int val=graph[curVal][i];
        
        if(isVisited[val]) continue;
        
        long long child = dfs(a, val);
        
        result+=abs(child);
        sum+=child;
        
    }
    
    return sum;
}

long long solution(vector<int> a, vector<vector<int>> edges) {
    
    bool isZero=true;
    long long v=0;
    
    for(auto val : a){
        if(val!=0) isZero=false;
        v+=val;
    }
    if(isZero) return 0;
    if(v!=0) return -1;
    
    init(edges);
    
    dfs(a,0); 
    
    return result;
}
